stdlib/join/equijoin.go

```diff
@@ -0,0 +1,169 @@
+package join
+
+import (
+	"context"
+
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/ast"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/influxdata/flux/interpreter"
+	"github.com/influxdata/flux/plan"
+	"github.com/influxdata/flux/semantic"
+)
+
+const EquiJoinKind = "equijoin"
+
+func init() {
+	plan.RegisterPhysicalRules(EquiJoinPredicateRule{})
+}
+
+type ColumnPair struct {
+	Left, Right string
+}
+
+type EquiJoinProcedureSpec struct {
+	On     []ColumnPair
+	As     interpreter.ResolvedFunction
+	Left   *flux.TableObject
+	Right  *flux.TableObject
+	Method string
+}
+
+func (p *EquiJoinProcedureSpec) Kind() plan.ProcedureKind {
+	return plan.ProcedureKind(EquiJoinKind)
+}
+
+func (p *EquiJoinProcedureSpec) Copy() plan.ProcedureSpec {
+	return &EquiJoinProcedureSpec{
+		On:     p.On,
+		As:     p.As,
+		Left:   p.Left,
+		Right:  p.Right,
+		Method: p.Method,
+	}
+}
+
+func (p *EquiJoinProcedureSpec) Cost(inStats []plan.Statistics) (cost plan.Cost, outStats plan.Statistics) {
+	return plan.Cost{}, plan.Statistics{}
+}
+
+func newEquiJoin(spec *JoinProcedureSpec, cols []ColumnPair) *EquiJoinProcedureSpec {
+	return &EquiJoinProcedureSpec{
+		On:     cols,
+		As:     spec.As,
+		Left:   spec.Left,
+		Right:  spec.Right,
+		Method: spec.Method,
+	}
+}
+
+type EquiJoinPredicateRule struct{}
+
+func (EquiJoinPredicateRule) Name() string {
+	return "equiJoinPredicate"
+}
+
+func (EquiJoinPredicateRule) Pattern() plan.Pattern {
+	return plan.Pat(Join2Kind, plan.Any(), plan.Any())
+}
+
+func (EquiJoinPredicateRule) Rewrite(ctx context.Context, n plan.Node) (plan.Node, bool, error) {
+	s := n.ProcedureSpec()
+	spec, ok := s.(*JoinProcedureSpec)
+	if !ok {
+		return nil, false, errors.New(codes.Internal, "invalid spec type on join node")
+	}
+
+	fnBody := spec.On.Fn.Block.Body
+
+	if len(fnBody) != 1 {
+		return nil, false, errors.New(codes.Invalid, "predicate should only contain a single return statement")
```

Comment:
Users might be confused by this error, since a valid lambda to get passed in might look something like `(l, r) => l.id == r.id and l.foo == r.bar` and doesn't need an explicit `return` statement. (I believe that the semantic graph will have a `Return` statement for that case, but it can be elided in the syntax)

Also, the message should note that there are some forms of lambda that are just not supported yet, so the user understands that they didn't do something wrong, it's just that we have more work to do.

Reply:
Ah, I see your point. I'm open to any suggestions you have for the wording. Maybe something like this?

```
some lambda expressions are not yet supported in join predicates - function body should only contain a single logical expression comparing equality of columns from the left and right tables
```

That seems kinda long winded, but I'm having trouble thinking of a more succinct way to communicate the problem.

Reply:
It would be easier if we could link them to some documentation explaining what is and isn't allowed in a predicate.

Reply:
Maybe once #4631 is finished, we can circle back and update these errors to just link to the generated fluxdoc page for the new join package.

Reply:
Sure, I like your message, you could also explicitly refer to the name of the parameter:
```
some expressions are not yet supported in the "on" parameter - function body should only contain a single logical expression comparing equality of columns from the left and right tables
```
(I guess the error message will have prefixed with something like `error in function join` or some such)